### PR TITLE
Server-side session enhancements

### PR DESCRIPTION
**Pull request by nextgens**
## What type of PR?

This is the backport of #1791 and #1787 targetted for 1.8

### CODE REVIEW

File: core/admin/mailu/configuration.py
Comment (ghostwheel42) [inline core/admin/mailu/configuration.py:59]: Hmm. I think I didn't document SESSION_KEY_BITS in docs/configuration.rst

Comment (nextgens) [reply core/admin/mailu/configuration.py:59]: this is a "manual" backport to 1.8 ... I suggest that we approve it unless something is terribly off... and keep working on improvements through the "normal" channel: in a different PR against master (that we can also backport as required but using mergify)

**Comment by Diman0** [comment]
All tests fail with the error:
  self._hmac    = hmac.new(hmac.digest(key, b'SESSION_UID_HASH', digest='sha256'), digestmod='sha256')
NameError: name 'hmac' is not defined

Hmac must be defined somewhere and this is not part of this PR but an earlier PR?
Ah yes. I see that in master it is defined in utils.py and __init__.py.
https://github.com/Mailu/Mailu/blob/4a5f6b1f921b3cf8862972c1cd7fb0b1ca1b65a4/core/admin/mailu/utils.py#L9
https://github.com/Mailu/Mailu/blob/4a5f6b1f921b3cf8862972c1cd7fb0b1ca1b65a4/core/admin/mailu/__init__.py#L9

Could you  add the declaration at the some locations in this PR?

stacktrace
```
 File "/usr/bin/flask", line 8, in <module>
    sys.exit(main())
  File "/usr/lib/python3.8/site-packages/flask/cli.py", line 894, in main
    cli.main(args=args, prog_name=name)
  File "/usr/lib/python3.8/site-packages/flask/cli.py", line 557, in main
    return super(FlaskGroup, self).main(*args, **kwargs)
  File "/usr/lib/python3.8/site-packages/click/core.py", line 717, in main
    rv = self.invoke(ctx)
  File "/usr/lib/python3.8/site-packages/click/core.py", line 1132, in invoke
    cmd_name, cmd, args = self.resolve_command(ctx, args)
  File "/usr/lib/python3.8/site-packages/click/core.py", line 1171, in resolve_command
    cmd = self.get_command(ctx, cmd_name)
  File "/usr/lib/python3.8/site-packages/flask/cli.py", line 515, in get_command
    rv = info.load_app().cli.get_command(ctx, name)
  File "/usr/lib/python3.8/site-packages/flask/cli.py", line 372, in load_app
    app = locate_app(self, import_name, name)
  File "/usr/lib/python3.8/site-packages/flask/cli.py", line 254, in locate_app
    return find_best_app(script_info, module)
  File "/usr/lib/python3.8/site-packages/flask/cli.py", line 76, in find_best_app
    app = call_factory(script_info, app_factory)
  File "/usr/lib/python3.8/site-packages/flask/cli.py", line 116, in call_factory
    return app_factory()
  File "/app/***/__init__.py", line 59, in create_app
    return create_app_from_config(config)
  File "/app/***/__init__.py", line 23, in create_app_from_config
    utils.session.init_app(app)
  File "/app/***/utils.py", line 426, in init_app
    app.session_config = MailuSessionConfig(app)
  File "/app/***/utils.py", line 264, in __init__
    self._hmac    = hmac.new(hmac.digest(key, b'SESSION_UID_HASH', digest='sha256'), digestmod='sha256')
NameError: name 'hmac' is not defined
```

**Comment by Diman0** [comment]
bors try

**Comment by nextgens** [comment]
> All tests fail with the error:
> self._hmac = hmac.new(hmac.digest(key, b'SESSION_UID_HASH', digest='sha256'), digestmod='sha256')
> NameError: name 'hmac' is not defined
> 
> Hmac must be defined somewhere and this is not part of this PR but an earlier PR?
> Ah yes. I see that in master it is defined in utils.py and **init**.py.
> 
> https://github.com/Mailu/Mailu/blob/4a5f6b1f921b3cf8862972c1cd7fb0b1ca1b65a4/core/admin/mailu/utils.py#L9
> 
> https://github.com/Mailu/Mailu/blob/4a5f6b1f921b3cf8862972c1cd7fb0b1ca1b65a4/core/admin/mailu/__init__.py#L9
> 
> Could you add the declaration at the some locations in this PR?

Done; sorry, I'm having a hard time with git and running the test suite locally here...

**Comment by nextgens** [comment]
bors try

**Comment by Diman0** [comment]
@nextgens @ghostwheel42 Everything is fine now with this backport pr?

**Comment by ghostwheel42** [comment]
I think it's okay but I didn't  test it yet. 
"import hmac" in __init__.py is not necessary I think. 
Everything else looks okay to me.

**Comment by nextgens** [comment]
As far as I'm concerned it's good to go and should get broader testing ASAP (merge it now that we know that people are running 1.8 ;) )

**Comment by Diman0** [comment]
I will run some smoke tests and will then approve the PR. I'm running 1.8 btw ;)

**Comment by Diman0** [comment]
Hmm. I receive an HTTP 500 error when logging in. Admin logs show that the hash is not recognized:
```
admin_1      | Traceback (most recent call last):
admin_1      |   File "/usr/lib/python3.8/site-packages/flask/app.py", line 2292, in wsgi_app
admin_1      |     response = self.full_dispatch_request()
admin_1      |   File "/usr/lib/python3.8/site-packages/flask/app.py", line 1815, in full_dispatch_request
admin_1      |     rv = self.handle_user_exception(e)
admin_1      |   File "/usr/lib/python3.8/site-packages/flask/app.py", line 1718, in handle_user_exception
admin_1      |     reraise(exc_type, exc_value, tb)
admin_1      |   File "/usr/lib/python3.8/site-packages/flask/_compat.py", line 35, in reraise
admin_1      |     raise value
admin_1      |   File "/usr/lib/python3.8/site-packages/flask/app.py", line 1813, in full_dispatch_request
admin_1      |     rv = self.dispatch_request()
admin_1      |   File "/usr/lib/python3.8/site-packages/flask/app.py", line 1799, in dispatch_request
admin_1      |     return self.view_functions[rule.endpoint](**req.view_args)
admin_1      |   File "/app/mailu/ui/views/base.py", line 18, in login
admin_1      |     user = models.User.login(form.email.data, form.pw.data)
admin_1      |   File "/app/mailu/models.py", line 426, in login
admin_1      |     return user if (user and user.enabled and user.check_password(password)) else None
admin_1      |   File "/app/mailu/models.py", line 381, in check_password
admin_1      |     result = context.verify(password, reference)
admin_1      |   File "/usr/lib/python3.8/site-packages/passlib/context.py", line 2338, in verify
admin_1      |     record = self._get_or_identify_record(hash, scheme, category)
admin_1      |   File "/usr/lib/python3.8/site-packages/passlib/context.py", line 2026, in _get_or_identify_record
admin_1      |     return self._identify_record(hash, category)
admin_1      |   File "/usr/lib/python3.8/site-packages/passlib/context.py", line 1131, in identify_record
admin_1      |     raise ValueError("hash could not be identified")
admin_1      | ValueError: hash could not be identified

```

This is on my test server. The SQLite database might contain data created by Mailu master branch ( I have tested many PRs on this server). That is probably the cause. 
I will test tomorrow again with a backup restored from my Mailu 1.8 production server and will report results again. I bet it will work with my 1.8 backup restored.

Please dont approve until I ran my extra test tomorrow.

**Comment by ghostwheel42** [comment]
The PR looks good to me.
The 1.8 branch still has some unused imports and 2 errors in unused code parts, but this is fixed in master and does not need to be backported.
We could catch the ValueError raised when the stored password hash is invalid (models.py line 381) and then log it and return False.

**Comment by Diman0** [comment]
The valueerror for the unknown hash should normally not occur. This indicates you used an unsupported hash. 
Testing with my restored 1.8 backup shows there are no issues. The master branch apparently uses a different hashing algorithm than branch 1.8. I must have tested this earlier on my test machine. 

Everything is fine. I will approve the PR.

**Review by Diman0 — APPROVED** [review]

**Pull request merged**
